tsdb/engine/tsm1/DESIGN.md

```diff
@@ -0,0 +1,411 @@
+
+Some issues with the existing structure is that seeking to a given timestamp for a key has a unknown cost.  This can cause variability in read performance that would very difficult to fix.  Another issue is that startup times for loading a TSM file would grow in proportion to number and size of TSM files on disk since we would need to scan the entire file to find all keys contained in the file.  This could be addressed by using a separate index like file or changing the index structure.
+
+We've chosen to update the block index structure to ensure a TSM file is fully self-contained, supports consistent IO characteristics for sequential and random accesses as well as provides an efficient load time regardless of file size.  The implications of these changes are that the index is slightly larger and we need to be able to search the index despite each entry being variably sized.
+
+The following are some alternative design options to handle the cases where the index is too large to fit in memory.  We are currently planning to use an indirect MMAP indexing approach for loaded TSM files.  
+
+### Indirect MMAP Indexing
+
+One option is to MMAP the index into memory and record the pointers to the start of each index entry in a slice.  When searching for a given key, the pointers are used to perform a binary search on the underlying mmap data.  When the matching key is found, the block entries can be loaded and search or a subsequent binary search on the blocks can be performed.
+
+A variation of this can also be done without MMAPs by seeking and reading in the file.  The underlying file cache will still be utilized in this approach as well.
+
+As an example, if we have an index structure in memory such as:
+
+ ```
+┌────────────────────────────────────────────────────────────────────┐
+│                               Index                                │
+├─┬──────────────────────┬──┬───────────────────────┬───┬────────────┘
+│0│                      │62│                       │145│
+├─┴───────┬─────────┬────┼──┴──────┬─────────┬──────┼───┴─────┬──────┐
+│Key 1 Len│   Key   │... │Key 2 Len│  Key 2  │ ...  │  Key 3  │ ...  │
+│ 2 bytes │ N bytes │    │ 2 bytes │ N bytes │      │ 2 bytes │      │
+└─────────┴─────────┴────┴─────────┴─────────┴──────┴─────────┴──────┘
+```
+
+We would build an `offsets` slices where each element pointers to the byte location for the first key in then index slice.
+
+```
+┌────────────────────────────────────────────────────────────────────┐
+│                              Offsets                               │
+├────┬────┬────┬─────────────────────────────────────────────────────┘
+│ 0  │ 62 │145 │
+└────┴────┴────┘
+ ```
+
+
+Using this offset slice we can find `Key 2` by doing a binary search over the offsets slice.  Instead of comparing the value in the offsets (e.g. `62`), we use that as an index into the underlying index to retrieve the key at postion `62` and perform our comparisons with that.
```

Comment:
Some various optimizations come to mind, like keeping, say, the first 4 bytes of each key with the offset -- it might allow a chunk of the search to place with hitting the disk at all. Something that might be worth keeping in mind.

It would double the RAM requirements, but it might be worth it.